### PR TITLE
Print usage to stderr rather than stdout to be more friendly to appli…

### PR DESCRIPTION
**Pull request by Happy0**
…cations that use unix pipelines.

### CODE REVIEW

**Comment by Happy0** [comment]
Hmm... This might not be the right thing to do, actually... What if a application has exposed a '--help' flag which calls exitWithUsage... In this case, it should probably print to stdout...

**Comment by Happy0** [comment]
I guess I should create the following new functions instead?:

exitWithUsageErr

exitWithUsageMessage and parseArgsOrExit will call this, rather than exitWithUsage as they are assumed to be called because of errors...

**Comment by ryanartecona** [comment]
Oh yeah, you're right. Good call.

**Comment by Happy0** [comment]
I've updated the pull request with the changes discussed above :P.

**Comment by ryanartecona** [comment]
Sorry for not taking a look quickly! Been busy at work.

The two `exitWithUsage{,Err}` seem a little out of line with `exitWithUsageMessage`, since (a) the former has 2 variants for stdout/stderr, and the latter doesn't, and (b) `exitWithUsageMessage` prints to stderr, even though that's not immediately obvious given the name (docs help, but it could be more obvious).

One idea would be to change them to `{exit,die}WithUsage` and `{exit,die}WithUsageMessage`, where the `die...` versions go to stderr (or 'die' could be 'fail', idk). If we go down that route, we could also rename `parseArgsOrExit` to `parseArgsOrDie`, without (?) keeping around a stdout variant. People will have to update their code with the renaming, but if we need a major version bump anyway, may as well, right?

Happy to make these changes on top of your PR if you agree, but I may not get to it immediately.

**Comment by Happy0** [comment]
No problem! Thanks for having a look.

Feel free to submit a pull request on the branch that my pull request is based on... Or merging and then making a new commit. Whatever you plan to do :P.

**Comment by Happy0** [comment]
@ryanartecona I had a spare 15 minutes on the train and I went ahead and implemented the changes you suggested. They make sense to me. Let me know if it looks alright to you.

Do you think I've bumped the version in the right column?

File: System/Console/Docopt/Public.hs [hunk 1]
```diff
@@ -47,32 +48,46 @@ import Data.Maybe (fromMaybe)
...
+import System.IO
 
 -- | Parse command line arguments.
 parseArgs :: Docopt -> [String] -> Either ParseError Arguments
 parseArgs parser = getArguments (optFormat parser)
 
--- | Same as 'parseArgs', but 'exitWithUsage' on parse failure. E.g.
+-- | Same as 'parseArgs', but 'dieWithUsage' on parse failure. E.g.
 --
 -- > args <- parseArgsOrExit patterns =<< getArgs
```
Comment (ryanartecona) [inline System/Console/Docopt/Public.hs]: Missed one!

File: System/Console/Docopt/Public.hs [hunk 2]
```diff
@@ -47,32 +48,46 @@ import Data.Maybe (fromMaybe)
...
 
--- | Exit after printing a custom message followed by usage text.
---   Intended for convenience when more context can be given about what went wrong.
+-- | Exit after printing a custom message followed by usage text to stdout.
+--   Intended for providing documentation to a user.
 exitWithUsageMessage :: Docopt -> String -> IO a
 exitWithUsageMessage doc msg = do
   putStrLn msg
-  putStrLn ""
+  putStrLn  ""
```
Comment (ryanartecona) [inline System/Console/Docopt/Public.hs]: This is a very minor nit, but now I can' t unsee it. :)

File: System/Console/Docopt/Public.hs [hunk 3]
```diff
@@ -47,32 +48,46 @@ import Data.Maybe (fromMaybe)
...
   exitWithUsage doc
 
+-- | Exit after printing a custom message followed by usage text to stderr
```
Comment (ryanartecona) [inline System/Console/Docopt/Public.hs]: Trail with a `.`

File: System/Console/Docopt/Public.hs [hunk 4]
```diff
@@ -47,32 +48,46 @@ import Data.Maybe (fromMaybe)
...
+-- | Exit after printing usage text to stdout.
 exitWithUsage :: Docopt -> IO a
 exitWithUsage doc = do
   putStr $ usage doc
-  exitFailure
+  exitSuccess
+
+-- | Exit after printing usage text to stderr.
+dieWithUsage :: Docopt -> IO a
+dieWithUsage doc = die $ usage doc
```
Comment (ryanartecona) [inline System/Console/Docopt/Public.hs]: I didn't notice before, but I'm pretty amused by the fact that `dieWithUsage doc` is now a _longer_ (in char count) invocation than would be `die $ usage doc`. This is a "composition FTW" if I've ever seen one.

File: System/Console/Docopt/Public.hs [hunk 5]
```diff
@@ -11,6 +11,7 @@ module System.Console.Docopt.Public
     , Docopt ()
     , usage
     , exitWithUsage
+    , dieWithUsage
```
Comment (ryanartecona) [inline System/Console/Docopt/Public.hs]: + `dieWithUsageMessage` ?

**Comment by ryanartecona** [comment]
Updates look great! :100: 

I think that's the right version component to bump. I don't think this lib is quite `1.0+`, but this is definitely a breaking change, so that looks right.

I made some inline notes. I expect to have some time this weekend to take care of them, make some other corresponding updates (examples, make sure the docs are all consistent, etc.), and I'll merge these in then. It will be a little while before I cut this release, since I want some other fixes to go out with it.

**Comment by Happy0** [comment]
\o/

I made changes based on your inline comments. I've also attempted to update the documentation examples - let me know if I've missed anything!

**Comment by ryanartecona** [comment]
Awesome.

Yeah, the README is the main one. There's also some stuff in `examples/` that will probably need a couple of those functions renamed before they're compilable again.

If you get to it, great; if not, no sweat. I need to get in here soon (again, hopefully this weekend) to see if I can address the other open issues, and if I can I will update the changelog and release this and those fixes in a new version. I'll leave this PR open until I get to those anyway.

**Comment by Happy0** [comment]
I updated the examples and made the function names in NoTH.hs (which I'd overlooked) consistent with the new conventions. Let me know if there's anything else :P.

**Comment by ryanartecona** [comment]
All set! Sorry for the slow back-and-forth, but thanks for the changes @Happy0!

**Comment by Happy0** [comment]
Hurray :D.

No problem at all. I've been busy as well! I'm glad we took the time to think about it and get it right :P.

**Pull request merged**